contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,133 @@
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    let connection_type = &invocation.connection_type;
+    let database_name = &invocation.db_name;
+    let request_guard_type = &invocation.type_name;
+    let pool_type = Ident::new(&format!("{}Pool", request_guard_type), request_guard_type.span());
+
+    // The Rocket config has to be cloned because otherwise without `#![feature(nll)]`, you end up
+    // with an error about not being able to move Rocket when adding to managed state or returning
+    // an `Err(rocket)` since it's borrowed by the `rocket.config()` method call.
+    let tokens = quote! {
+        pub struct #request_guard_type(
+            pub ::rocket_contrib::databases::r2d2::PooledConnection<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>
+        );
+        pub struct #pool_type(::rocket_contrib::databases::r2d2::Pool<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+
+        impl #request_guard_type {
+            pub fn fairing() -> impl ::rocket::fairing::Fairing {
+                use ::rocket_contrib::databases::Poolable;
+
+                ::rocket::fairing::AdHoc::on_attach(|rocket| {
+                    let pool = ::rocket_contrib::databases::database_config(#database_name, rocket.config())
+                        .map(#connection_type::pool);
+
+                    match pool {
+                        Ok(Ok(p)) => Ok(rocket.manage(#pool_type(p))),
+                        Err(config_error) => {
+                            config_error.pretty_print(#database_name);
+                            Err(rocket)
+                        },
+                        Ok(Err(pool_error)) => {
+                            ::rocket_contrib::log_err(&format!("Error initializing pool for '{}': {:?}", #database_name, pool_error));
+                            Err(rocket)
+                        },
+                    }
+                })
+            }
+
+            /// Retrieves a connection of type `Self` from the `rocket` instance. Returns `Some` as long as
+            /// `Self::fairing()` has been attached and there is at least one connection in the pool.
+            pub fn get_one(rocket: &::rocket::Rocket) -> Option<Self> {
+                rocket.state::<#pool_type>().and_then(|p| { // Not terribly enthused about `and_then`, but I can't use `?` for this one
```

Comment:
Why can't you use `map`?

```rust
rocket.state::<#pool_type>()
    .and_then(|pool| pool.0.get().ok())
    .map(#request_guard_type)
```

Reply:
I can! I don't know why I missed that... Good catch!